### PR TITLE
Strategy should work with 1 replica

### LINKED ISSUE #7: Strategy checks don't account for "rounding up" pod counts
This breaks strategy progression when we're looking at small pod counts: it's easy for the incumbent to be stable with too many pods.

### PR DESCRIPTION
**Pull request by isutton**
This patch makes sure that the strategy works with one replica as well as many.

The intended behavior was to not complete the strategy right away but march through the strategy specification even if the whole transition capacity has been fulfilled.

Closes #7.

### CODE REVIEW

**Comment by kanatohodets** [comment]
I rebased this branch on master to get the TravisCI pipeline engaged.

File: pkg/controller/strategy/comparisons.go [hunk 1]
```diff
@@ -1,11 +1,23 @@
 package strategy
 
-func contenderCapacityComparison(achieved uint, desired uint) bool {
-	return achieved >= desired
+func contenderCapacityComparison(achievedPercentage, desiredPercentage, totalNumberOfPods uint) bool {
+	// Consider that contender capacity has achieved its desired state if
```
Comment (kanatohodets) [inline pkg/controller/strategy/comparisons.go]: I think it might make sense to explain here and in the `incumbentCapacityComparison` about our "pod math", where we always round up to the nearest whole pod (and so 10% desired capacity of 1 pod is 1 pod).

File: pkg/controller/strategy/controller_test.go [hunk 2]
```diff
@@ -30,125 +30,142 @@ func init() {
 }
 
 func TestContenderReleasePhaseIsWaitingForCommandForInitialStepState(t *testing.T) {
-	f := newFixture(t)
+	for _, i := range []uint{1, 10} {
```
Comment (kanatohodets) [inline pkg/controller/strategy/controller_test.go]: How would you feel about having a global slice with the pod counts to test that all the test cases share? That way we can put a comment there explaining why we think these specific values should be tested (mentioning the "pod math" bits).

File: test/e2e/e2e_test.go [hunk 3]
```diff
@@ -268,13 +267,20 @@ func TestNewApplicationVanguard(t *testing.T) {
 	}
 }
 
-func TestRolloutVanguard(t *testing.T) {
+func TestNewApplicationVanguardMultipleReplicas(t *testing.T) {
+	testNewApplicationVanguard(4, t)
```
Comment (kanatohodets) [inline test/e2e/e2e_test.go]: Also probably worth putting a note as to why we have these distinct cases (1 pod vs 4 pods).

File: pkg/controller/strategy/comparisons.go [hunk 4]
```diff
@@ -1,11 +1,23 @@
...
+	return achievedPercentage >= desiredPercentage
 }
 
-func incumbentCapacityComparison(achieved uint, desired uint) bool {
-	return achieved <= desired
+func incumbentCapacityComparison(achievedPercentage, desiredPercentage, totalNumberOfPods uint) bool {
+	// Consider that the incumbent capacity has achieved its desired state if
+	// the end game is one pod and we already have it, as long as the desired
+	// is smaller than 100.
+	if totalNumberOfPods == 1 && achievedPercentage == 100 && desiredPercentage < 100 {
```
Comment (kanatohodets) [inline pkg/controller/strategy/comparisons.go]: I don't totally understand the  `desired < 100` check; my first impression is that it is a bug, since it might say its OK to proceed when we should actually be scaling down to 0. 

Reasoning: if `desiredPercentage` is 0 on a 1 pod deployment, and we (for whatever reason) haven't made any progress turning off that pod, I think this will say that everything is correct, when really it should say that we're not there yet. I think the correct test might be: `totalPods==1 && achievedPercent == 100 && desiredPercentage > 0` for the incumbent as well.

File: pkg/controller/errors.go [hunk 5]
```diff
@@ -6,28 +6,36 @@ import (
 	"k8s.io/apimachinery/pkg/types"
 )
 
-type MultipleOwnerReferencesError error
+type MultipleOwnerReferencesError string
```
Comment (kanatohodets) [inline pkg/controller/errors.go]: Is this kind of a drive-by fix for this PR, or was it required to make things work? It doesn't seem related to the rest at first glance (our wide-spread bugs with error types notwithstanding).

Comment (isutton) [reply pkg/controller/errors.go]: Yes, it was a drive-by, but to fix the bug where `shouldRetry` is always evaluated to `false` (see below).

https://github.com/bookingcom/shipper/blob/3015272fe0ec4b28bf67fdad0279d8e6f4e120ec/pkg/controller/installation/installation_controller.go#L187-L190

I'm fine to revert the related changes and create another PR for it.

Comment (kanatohodets) [reply pkg/controller/errors.go]: Ok cool, no need to split it. I'll grab this in the structured errors work either way.

**Comment by kanatohodets** [comment]
Reflecting our chat today: we understand now that the bug is with any odd numbered replica count on the incumbent. 

Reasoning: if the incumbent has 5 pods, and we are trying to scale it to 50% of its full capacity (2.5 pods), we will scale it to 3 pods (2.5 rounded up to nearest whole pod). As a result, the capacity target object will report that the achieved capacity is 3/5, which is larger than what the strategy thinks the incumbent should be at (2.5), and the rollout gets stuck waiting for the incumbent to scale down.

File: pkg/chart/cache/cache.go [hunk 6]
```diff
@@ -121,6 +123,7 @@ func clean(names ...string) (string, string, string) {
 		// names a nicer read (not http:__blorg.baz.com).
 		names[i] = strings.TrimPrefix(names[i], "https://")
 		names[i] = strings.TrimPrefix(names[i], "http://")
+		names[i] = strings.Replace(names[i], ":", "_", -1)
```
Comment (ksurent) [inline pkg/chart/cache/cache.go]: I'm assuming these are windows–specific?

File: pkg/controller/traffic/pod_label_shifter.go [hunk 7]
```diff
@@ -217,13 +218,15 @@ func calculateReleasePodTarget(releasePods int, releaseWeight uint32, totalPods
 	if totalWeight == 0 {
 		targetPercent = 0
 	} else {
-		targetPercent = float64(releaseWeight) / float64(totalWeight)
+		targetPercent = float64(releaseWeight) / float64(totalWeight) * 100
 	}
 	// Round up to the nearest pod, clamped to the number of pods this release has.
-	targetPods := int(
+	targetPods := int(replicas.CalculateDesiredReplicaCount(uint(totalPods), float64(targetPercent)))
```
Comment (ksurent) [inline pkg/controller/traffic/pod_label_shifter.go]: `targetPercent` is already defined as `float64`.

File: pkg/chart/cache/cache.go (hunk 6 above)
Comment (ksurent) [reply pkg/chart/cache/cache.go]: Nvm, found it.
